cdap-kms/src/main/java/co/cask/cdap/security/store/KMSSecureStore.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.security.store;
+
+import co.cask.cdap.api.security.store.SecureStore;
+import co.cask.cdap.api.security.store.SecureStoreData;
+import co.cask.cdap.api.security.store.SecureStoreManager;
+import co.cask.cdap.api.security.store.SecureStoreMetadata;
+import com.google.inject.Inject;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.crypto.key.KeyProvider;
+import org.apache.hadoop.crypto.key.KeyProviderFactory;
+import org.apache.hadoop.crypto.key.kms.KMSClientProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Secure Store implementation backed by Hadoop KMS. This class is loaded using reflection if
+ * the provider is set to kms and Hadoop version is 2.6.0 or higher.
+ */
+@SuppressWarnings("unused")
+class KMSSecureStore extends AbstractSecureStore implements SecureStore, SecureStoreManager {
+  private static final Logger LOG = LoggerFactory.getLogger(KMSSecureStore.class);
+  /**
+   * Hadoop KeyProvider interface. This is used to interact with KMS.
+   */
+  private KeyProvider provider;
+  private final Configuration conf;
+
+  /**
+   * Sets up the key provider.
+   * @param conf Hadoop configuration. core-site.xml contains the key provider URI.
+   * @throws URISyntaxException If the key provider path is not a valid URI.
+   * @throws IOException If the authority or the port could not be read from the provider URI.
+   */
+  @Inject
+  KMSSecureStore(Configuration conf) throws IOException, URISyntaxException {
+    this.conf = conf;
+    try {
+      URI providerUri = new URI(conf.get(KeyProviderFactory.KEY_PROVIDER_PATH));
+      provider = KMSClientProvider.Factory.get(providerUri, conf);
+    } catch (URISyntaxException e) {
+      throw new URISyntaxException("Secure store could not be loaded. The value for hadoop.security.key.provider.path" +
```

Comment:
thinking about this again - what are the implications of not bailing in the constructor, and instead initializing the provider lazily when required? Like I said this may get buried under a guice stacktrace and could be difficult to debug